collector/pressure_linux.go

```diff
@@ -0,0 +1,123 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !nopressure
+
+package collector
+
+import (
+	"fmt"
+	"os"
+	"syscall"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/procfs"
+)
+
+var (
+	psiResources = []string{"cpu", "io", "memory"}
+	psiErrPrefix = "pressure information is not available"
+)
+
+type pressureStatsCollector struct {
+	cpu     *prometheus.Desc
+	io      *prometheus.Desc
+	ioFull  *prometheus.Desc
+	mem     *prometheus.Desc
+	memFull *prometheus.Desc
+
+	fs procfs.FS
+}
+
+func init() {
+	registerCollector("pressure", defaultEnabled, NewPressureStatsCollector)
+}
+
+// NewPressureStatsCollector returns a Collector exposing pressure stall information
+func NewPressureStatsCollector() (Collector, error) {
+	fs, err := procfs.NewFS(*procPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open procfs: %v", err)
+	}
+
+	return &pressureStatsCollector{
+		cpu: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "pressure", "cpu_waiting_seconds_total"),
+			"Total time in seconds that processes have waited for CPU time",
+			nil, nil,
+		),
+		io: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "pressure", "io_waiting_seconds_total"),
+			"Total time in seconds that processes have waited due to IO congestion",
+			nil, nil,
+		),
+		ioFull: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "pressure", "io_stalled_seconds_total"),
+			"Total time in seconds no process could make progress due to IO congestion",
+			nil, nil,
+		),
+		mem: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "pressure", "memory_waiting_seconds_total"),
+			"Total time in seconds that processes have waited for memory",
+			nil, nil,
+		),
+		memFull: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "pressure", "memory_stalled_seconds_total"),
+			"Total time in seconds no process could make progress due to memory congestion",
+			nil, nil,
+		),
+		fs: fs,
+	}, nil
+}
+
+// Update calls procfs.NewPSIStatsForResource for the different resources and updates the values
+func (c *pressureStatsCollector) Update(ch chan<- prometheus.Metric) error {
+	pth := c.fs.Path("pressure")
+	fi, err := os.Stat(pth)
+	if err != nil {
+		if e, ok := err.(*os.PathError); ok && e.Err == syscall.ENOENT {
+			log.Debugf("%s; requires a kernel with CONFIG_PSI", psiErrPrefix)
+			return nil
+		}
+		log.Debugf("%s: %v", psiErrPrefix, err)
+		return nil
+	}
+	if !fi.IsDir() {
+		log.Debugf("%s: %s must be a directory", psiErrPrefix, pth)
+		return nil
+	}
+
```

Comment:
This prereq block isn't sitting well with me. Ideally I would move this into `NewPressureStatsCollector` but:
* any error returned in NewPressureStatsCollector causes the node exporter to terminate
* if I log but don't return an error we end up returning `nil nil` which causes problems b/c a nil error signifies it should be safe to use the returned collector and call `Update()` on it

Reply:
I think it would be better to return the error from `Update()` instead of logging to debug and returning nil.  Both in the prereq checks and when the stats are collected.  That way it will be obvious to the user that the pressure collector is failing and they can disable it.  Otherwise it will be reporting success even though it's not collecting anything.

It would probably be ok to leave out the prereq checks, as long as you are returning the error from procfs.  Or if you want to only do the prereq check once at startup maybe you could use [sync.Once](https://golang.org/pkg/sync/#Once) or just a boolean flag.

Reply:
tbh, I don't really have a preference between:

- returning error from Update()
- removing prereq check
- doing them in NewPressureStatsCollector and disabling the collector by default

All seem reasonble. Agree with @pgier that just logging and returning nil is just not a good idea. Leaving this up to @daenney

Reply:
Alright, one more to consider. In theory `NewPSIStatForResource` could fail for `memory`, but pass for `cpu`, or at some point a new resource could be added that only works on newer kernels. If we immediately return on error we might end up failing to collect some metrics. I'm considering changing the error handling in such a way that it accumulates the errors and returns a single one at the end of `Update()`. Does that seem OK?

Reply:
Failing for metrics is fine, we prefer to have whole collectors fail in the case of errors.

For example, this is why we separated the `cpu` and `cpufreq` collectors.

For features existing in some kernels but not others, we typically have the procfs functions return pointers to data, in the case of "optional" features, we return `nil` as long as we had no other error than "feature not implemented". See the nfsd metrics for examples.

Reply:
Based on what you're saying I think we should change the procfs implementation first, over [here](https://github.com/prometheus/procfs/blob/master/proc_psi.go#L80-L82).

Currently if it can't open the file, which I think qualifies as "a feature existing in some kernels but not others", it returns an error. Based on what you're saying we should return a `nil` there instead of the current error?

It's a bit annoying in the case of PSI. As it currently stands, memory, cpu and io should always work if you have a kernel with `CONFIG_PSI`, they're the original 3 resources introduced. It's possible, albeit unlikely, that more resource types could be introduced in the future which would effectively be "optional" features. This is what I'm trying to protect a bit against as I think it would be counter-intuitive that if a feature were added to PSI it would stop the collector from being able to collect the current set of data.

Reply:
Let's not overthink this. To move this forward, let's just go with the simplest option and remove the revisit and return error on first error. If there will be a kernel with a different set where this becomes a problem, we can revisit it.

Reply:
Sounds good. I'll go for that.